sycl/doc/extensions/experimental/sycl_ext_oneapi_usm_prepare_release_device_copy.asciidoc

```diff
@@ -0,0 +1,99 @@
+of The Khronos Group Inc.  OpenCL(TM) is a trademark of Apple Inc. used by
+permission by Khronos.
+
+== Contact
+
+To report problems with this extension, please open a new issue at:
+
+https://github.com/intel/llvm/issues
+
+
+== Dependencies
+
+This extension is written against the SYCL 2020 revision 5 specification.  All
+references below to the "core SYCL specification" or to section numbers in the
+SYCL specification refer to that revision.
+
+== Status
+
+This is an experimental extension specification, intended to provide early
+access to features and gather community feedback. Interfaces defined in this
+specification are implemented in DPC\++ but they are not finalized and may
+change incompatibly in future versions of DPC++ without prior notice.
+Shipping software products should not rely on APIs defined in this
+specification.
+
+
+== Overview
+
+Data transfer between Host and Device is a common operation when using SYCL. Data transfers may be explicit, using the SYCL memcpy functions, or implicit via use of SYCL buffers and accessors.
+When these transfers are between USM memory and USM memory the transfer rate is maximized. When either source or destination is non-USM, data transfer is slower.
+
+The Level Zero APIs ```zexDriverImportExternalPointer``` and ```zexDriverReleaseImportedPointer``` are available to temporarily promote system memory to USM. The import/release can be done for the duration that the host pointer is a source/destination of host/device data transfers without needing to change the original allocation.
+
+This extension supports the Level Zero USM import/release functionality through SYCL interfaces.
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine
+the macro `SYCL_EXT_ONEAPI_USM_IMPORT_RELEASE` to one of the values defined
+in the table below.  Applications can test for the existence of this macro
+to determine if the implementation supports this feature, or applications
+can test the macro's value to determine which of the extension's features
+the implementation supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|Initial version of this extension.
+|===
+
+=== API of the extension
+
+SYCL USM import/release APIs give the programmer explicit control over the range and duration of imports. The APIs are analogous to the ```malloc_host``` APIs:
+
+```c++
+namespace sycl::ext::oneapi::experimental {
+
+  void prepare_for_device_copy(void *Ptr, size_t NumBytes,
+                               const context &Context);
+  void prepare_for_device_copy(void *Ptr, size_t NumBytes,
+                               const queue &Queue);
+
+  void release_from_device_copy(void *Ptr, const context &Context);
+  void release_from_device_copy(void *Ptr, const queue &Queue);
+
+} // namespace sycl::ext::oneapi::experimental
+```
```

Comment:
You should follow this synopsis with a table describing the functions.  Here's a start, but the description should be expanded to answer the questions I have below:

> `prepare_for_device_copy(void *Ptr, size_t NumBytes, const context &Context)`
>
> Informs the implementation that the host memory range starting at `Ptr` and extending for `NumBytes` bytes may be used as either the source or destination of an explicit copy operation to a queue using context `Context`.  Calling this function may allow the implementation to accelerate these copy operations.  Generally, it is only useful to call this function if the host memory range will be used for several subsequent copy operations; it is not generally useful to call it before each individual copy operation.
>
> `prepare_for_device_copy(void *Ptr, size_t NumBytes, const queue &Queue)`
>
> Same as `prepare_for_device_copy(Ptr, NumBytes, Queue.get_context())`.
>
> `release_from_device_copy(void *Ptr, const context &Context)`
>
> Undoes the effect of a previous call to `prepare_for_device_copy` on `Ptr`.  It is still valid to copy to or from `Ptr`, but these copies might not be optimized.
>
> `release_from_device_copy(void *Ptr, const queue &Queue)`
>
> Same as `release_from_device_copy(Ptr, Queue.get_context())`.

Some questions I have:

* Is it valid to call `prepare_for_device_copy` on the same host memory range for two different contexts?
* What happens if the `Ptr` passed to `release_from_device_copy` does not match the value passed to a previous call to `prepare_for_device_copy`?  Is this silently ignored?  Is it an error which causes the runtime to throw an exception?
* What happens if there are nested calls to the same host memory range such as:
  ```
  prepare_for_device_copy(ptr, size, ctxt);
  prepare_for_device_copy(ptr, size, ctxt);
  /* ... */
  release_from_device_copy(ptr, ctxt);
  release_from_device_copy(ptr, ctxt);
  ```
  Is this an error?  Does the first call to `release_from_device_copy` undo the effect of *both* previous calls to `prepare_for_device_copy`?  Does the runtime keep a "use count"?